packages/story-editor/src/components/canvas/useInsertTextSet.js

```diff
@@ -44,24 +49,142 @@ function useInsertTextSet() {
   );
 
   const insertTextSet = useBatchingCallback(
-    (toAdd) => {
+    async (toAdd) => {
+      const htmlFormatters = getHTMLFormatters();
+      const { setColor } = htmlFormatters;
       const addedElements = [];
-      toAdd.forEach((element) => {
+
+      const hasPredefinedColor = toAdd.some((element) =>
+        element.content?.includes('color:')
+      );
+
+      let textElementsContrasts = [];
+      let firstValidColor = null;
+
+      const enableSmartTextColor = true;
+      const white = {
+        r: 255,
+        g: 255,
+        b: 255,
+      };
+      const black = {
+        r: 0,
+        g: 0,
+        b: 0,
+      };
+      const whiteScrim = {
+        ...white,
+        a: 0.3,
+      };
+      const blackScrim = {
+        ...black,
+        a: 0.3,
+      };
+      let preferredScrimColor, scrimsCount, useScrim;
+
+      // Insert scrim as a first element if needed.
+      if (enableSmartTextColor && !hasPredefinedColor) {
+        textElementsContrasts = await Promise.all(
+          toAdd.map((element) =>
+            element.type === 'text'
+              ? calculateAccessibleTextColors(element)
+              : null
+          )
+        );
+        firstValidColor = textElementsContrasts.find((c) => c);
+        const allColorsEqual = textElementsContrasts.every(
+          (contrast) =>
+            contrast === null ||
+            (contrast.color.r === firstValidColor.color.r &&
+              contrast.color.g === firstValidColor.color.g &&
+              contrast.color.b === firstValidColor.color.b &&
+              contrast.color.a === firstValidColor.color.a)
+        );
+        scrimsCount = textElementsContrasts.reduce(
+          (acc, contrast) => (contrast?.backgroundColor ? acc + 1 : acc),
+          0
+        );
+        const blackScrims = textElementsContrasts.reduce(
+          (acc, contrast) =>
+            contrast?.backgroundColor?.r === 0 &&
+            contrast?.backgroundColor?.g === 0 &&
+            contrast?.backgroundColor?.b === 0
+              ? acc + 1
+              : acc,
+          0
+        );
+        preferredScrimColor =
+          scrimsCount - blackScrims > 0 ? blackScrim : whiteScrim;
+
+        useScrim = scrimsCount > 0 || allColorsEqual === false;
+
+        if (useScrim) {
+          const { textSetHeight, textSetWidth } = toAdd[0];
+          const textSetXY = toAdd.reduce(
+            (bb, e) => {
+              if (e.x < bb.x || !bb.x) {
+                bb.x = e.x;
+              }
+              if (e.y < bb.y || !bb.y) {
+                bb.y = e.y;
+              }
+
+              return bb;
+            },
+            { x: null, y: null }
+          );
+          const { x, y } = textSetXY;
```

Comment:
You can just map the values out and pass the whole array to `Math.min`:

```suggestion
          const getMinDim = (dim) => Math.min.apply(null, toAdd.map(e=>e[dim]));
          const x = getMinDim('x');
          const y = getMinDim('y');
```